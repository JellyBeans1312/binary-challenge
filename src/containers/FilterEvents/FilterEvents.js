import React, { Component } from 'react';
import { Event } from '../../components/Event/Event';
import { connect } from 'react-redux';
import './FilterEvents.css'

export class FilterEvents extends Component {
  constructor() {
    super();
    this.state = {
      filteredEvents: [],
      filterButton: "",
      filters: []
    };
  }

  componentDidMount() {
    this.gatherFilters();
  }

  setFilters = e => {
    let selected = e.target.value;
    let selectedEvents = this.props.events.filter(
      event => event.category === selected
    );
    this.setState({ filteredEvents: selectedEvents });
  };

  genreFilter = () => {
    return this.state.filteredEvents.map(event => {
      return (
        <div className="event-container">
          <Event
            {...event}
            key={event.id}
          />
        </div>
      );
    });
  };

  gatherFilters = () => {
    let filters = [];
    this.props.events.forEach(event => {
      if (!filters.includes(event.category)) {
        filters.push(event.category);
      }
    });
    this.setState({ filters });
  };

  populateGenreButtons = () => {
    return this.state.filters.map(category => {
      return (
        <>
          <label htmlFor={category}>{category}</label>
          <input
            type="radio"
            name="genre"
            value={category}
            placeholder="radio buttons"
            onClick={this.setFilters}
          ></input>
        </>
      );
    });
  };

  generateEvents = () => {
    return this.props.events.map(event => {
      return (
        <div className="event-container">
          <Event
            {...event}
            key={event.id}
          />
        </div>
      );
    });
  };

  clearInputs = () => {
    this.generateEvents()
    this.setState({ filteredEvents: [], filterButton: '' })
  }

  render() {
    return (
      <section>
        <div>
          <fieldset>
            <h3>Filter By Genre</h3>
            <label htmlFor="show all">Show All</label>
            <input
              type="radio"
              name="genre"
              value="show all"
              placeholder="radio buttons"
              onClick={this.clearInputs}
            ></input>
            {this.populateGenreButtons()}
          </fieldset>
        </div>
        <div className="event-container">
          {this.state.filteredEvents.length === 0 && this.generateEvents()}
          {this.state.filteredEvents.length > 0 && this.genreFilter()}
        </div>
      </section>
    );
  }
}

export const mapStateToProps = state => ({
  events: state.events
})

export default connect(mapStateToProps)(FilterEvents)