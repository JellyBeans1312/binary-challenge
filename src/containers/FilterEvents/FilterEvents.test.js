import React from "react";
import { shallow } from "enzyme";
import { FilterEvents, mapStateToProps }from "./FilterEvents";

describe("FilterEvents ", () => {
  let wrapper;
  const mockEvents = [
    { title: "something", category: 'concert', location: [{lat: 23, long: 234}] }
  ];

  beforeEach(() => {
    wrapper = shallow(
      <FilterEvents
        events={mockEvents}
      />
    );
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should filter events based on value of filter and add event to state", () => {
    const mockEvent = { target: { value: "concert" } };
    expect(wrapper.state("filteredEvents")).toEqual([]);
    wrapper.find("input").at(1).simulate("click", mockEvent);
    expect(wrapper.state("filteredEvents")).toEqual(mockEvents);
  });

  it("should gather filters based on event category", () => {
    wrapper.instance().gatherFilters();
    expect(wrapper.state("filters")).toEqual(["concert"]);
  });

  it("should reset state on click", () => {
    wrapper.find("input").at(0).simulate("click");

    expect(wrapper.state("filteredEvents")).toEqual([]);
    expect(wrapper.state("filterButton")).toEqual("");
  });

  describe('mapStateToProps', () => {
    it('should return an array of events', () => {
      const mockState = {
        events: [{ title: "mirrorland"}]
      };
      const expected = {
        events: [{ title: "mirrorland"}]
      };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
    });
  });