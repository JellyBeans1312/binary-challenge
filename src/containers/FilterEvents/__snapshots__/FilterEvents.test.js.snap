// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterEvents  should match snapshot 1`] = `
<section>
  <div>
    <fieldset>
      <h3>
        Filter By Genre
      </h3>
      <label
        htmlFor="show all"
      >
        Show All
      </label>
      <input
        name="genre"
        onClick={[Function]}
        placeholder="radio buttons"
        type="radio"
        value="show all"
      />
      <label
        htmlFor="concert"
      >
        concert
      </label>
      <input
        name="genre"
        onClick={[Function]}
        placeholder="radio buttons"
        type="radio"
        value="concert"
      />
    </fieldset>
  </div>
  <div
    className="event-container"
  >
    <div
      className="event-container"
    >
      <Event
        category="concert"
        location={
          Array [
            Object {
              "lat": 23,
              "long": 234,
            },
          ]
        }
        title="something"
      />
    </div>
    <div
      className="event-container"
    >
      <Event
        category="concert"
        location={
          Array [
            Object {
              "lat": 234,
              "long": 2345,
            },
          ]
        }
        title="else"
      />
    </div>
  </div>
</section>
`;
